### PR TITLE
RELEASE 2.0.0

### PR DESCRIPTION
**Pull request by madpah**
Lil hand up for @madpah , just sort of the way towards having Services implemented (but not all the way there).

This PR will contain a number of additional features:
- #123 
- #155
- #149 
- #143
- #124
- #159 
- #6 
- #163
- #147 
- #171 
- #170 

Bugs squashed:
- #150
- #151
- #153
- #169 
- #176
- #179

This WIP PR is being squashed into a new feature branch where work will continue.

Note: Work originally squashed from #145.

supersedes #154

### CODE REVIEW

**Comment by jkowalleck** [comment]
@madpah let me know if you need a second opinion or a code review

File: cyclonedx/exception/output.py
Comment (jkowalleck) [inline cyclonedx/exception/output.py:25]: was this a public class? if so, the removal will cause a breaking change for some people downstream

File: cyclonedx/model/__init__.py
Comment (jkowalleck) [inline cyclonedx/model/__init__.py:77]: whaw can both be falsy? `flow` is an Enum, right? so it is never falsy.

File: cyclonedx/model/__init__.py
Comment (jkowalleck) [inline cyclonedx/model/__init__.py:797]: good idea.
alternatively the `repr` could be simply the `self.text`.

File: cyclonedx/model/bom.py
Comment (jkowalleck) [inline cyclonedx/model/bom.py:153]: please use enforce kwargs so the class is kind of future-prove
```path
- def __init__(self, components: Optional[List[Component]] = None, services: Optional[List[Service]] = None,
-                   external_references: Optional[List[ExternalReference]] = None) -> None:
+ def __init__(self, *, 
+                    components: Optional[List[Component]] = None, services: Optional[List[Service]] = None,
+                    external_references: Optional[List[ExternalReference]] = None) -> None:
```

File: cyclonedx/model/bom.py
Comment (jkowalleck) [inline cyclonedx/model/bom.py:198]: this is a breaking change

File: cyclonedx/model/bom.py
Comment (jkowalleck) [inline cyclonedx/model/bom.py:308]: this would not be needed, if `self._services` was a list at all times and would be returned by the setter.
then the users could run `myBom.services.append()`

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: cyclonedx/exception/output.py
Comment (madpah) [reply cyclonedx/exception/output.py:25]: It was a public class @jkowalleck. Agree this is a breaking change.

File: cyclonedx/model/component.py
Comment (jkowalleck) [inline cyclonedx/model/component.py:136]: no need to call `hash()` on each element in the tuple constructor.
the tuple's implementation of `hash` is expected to call `hash` in the tuples elements individually, if needed.
this means: just drop in all elements ala `hash((self.uid, self.url, self.author, self.committer, self.message))`

sample code for testing:
```py
class C:
    def __init__(self,a,b):
        self.a = a 
        self.b = b 

    def __hash__(self):
        h = hash((self.a, self.b))
        print( 'hash of ', id(self), repr(self), repr(self.a), repr(self.b), 'is', h)
        return h

i = C(C('foo', None),'bar')

print(hash(i))
```

File: cyclonedx/model/component.py
Comment (jkowalleck) [inline cyclonedx/model/component.py:153]: why the underscore in `copytight_`? 
is there another variable called `copyright`? i assume there is a module with that name.

File: cyclonedx/model/component.py
Comment (jkowalleck) [inline cyclonedx/model/component.py:464]: again: if the internal property was a list at all times, 
this mehod would not be needed, because people could call `object.descendants.append(foo)` to modify internal ata structures.
lass code, mor clear data structires.

File: cyclonedx/model/bom.py
Comment (madpah) [reply cyclonedx/model/bom.py:153]: Have adopted PEP-3102 now as we previously discussed in all model classes.

File: cyclonedx/model/issue.py
Comment (jkowalleck) [inline cyclonedx/model/issue.py:286]: no need for the list-constructor `[]` inside the tuple constructor `tuple()`
the tuple constructor accepts iterable as an constructor property.

and then the `set` sinstructor around the `sorted` is not needed. `sorted` returns a pre-sorted object, `set` does not support orders at all.

and then the `hash()` call is not needed, as the outer `hash()` will call `hash()` on the inner elements if needed.

the following should be enough
```patch
- tuple([hash(ref) for ref in set(sorted(self._references, key=hash))])
+ tuple(sorted(self._references, key=hash))
```

when looking at the code i see a huge design opportunity: 
all the element lists (like `references` etc ...) are actually intended to be not `list`but `set` in the first place.
from the spec the order of the elements does not matter at all. 

:1st_place_medal:  conclusion: 
* all the optional lists should be  sets. 
  *  therefore the internal property would be a set at all times
  * the getter would return `set` at all times
  * the parameters in the getter and constructor would accept not just sets but all implementations of `Iterable`
* do not call unnecessary `hash()` on internal implementations

**Comment by jkowalleck** [comment]
@madpah a huge change i would suggest is: https://github.com/CycloneDX/cyclonedx-python-lib/pull/148/files#r798309225

this way a lot of problems would be solved: 
* no need for implementing `object.add_<list-property>()` all the time
* the order of the list-properties would no longer matter - no sorting, no whatsoever
* serializers/normalizers would always have a reliable non-optional data type to work with (remeber: an empty `set` is falsy - for easy size-checks)
* additionally: use the `set`-literal: instead of writing `set([...])` simply write `{...}` - if this was needed at any place, still

File: cyclonedx/model/component.py
Comment (madpah) [reply cyclonedx/model/component.py:153]: `copyright` is a keyword and I was following [PEP-0008](https://www.python.org/dev/peps/pep-0008/#id42)

File: cyclonedx/model/issue.py
Comment (madpah) [reply cyclonedx/model/issue.py:286]: Agree with both points.

I'll initially address the `hash()` calls in this PR.

Have already raised #159 regarding use of `Set` vs `List`. Will be a big code change and just weighing up time and whether to hold back `2.0.0` for this or not?

Comment (madpah) [reply cyclonedx/model/issue.py:286]: Gah - simplifying the hash code as you call out basically requires us to migrate to Sets:

```
    tuple(sorted(self.external_references, key=hash)), tuple(sorted(self.properties, key=hash)),
TypeError: 'NoneType' object is not iterable
```

File: cyclonedx/model/component.py
Comment (jkowalleck) [inline cyclonedx/model/component.py:769]: seeing this change makes me wonder, if it might be a goot sime to switch the `bom_ref` from optional strings to 
an non-optional read-only object of a new data type `BomRef`.

the ref/bom-ref properties in CycloneDX are a special kind of string. they are inteded to link elements together.
therefore bom-refs are used, instead of the actual linked elements.
having special data types makes it easy to tell elements apart by the reference.

proposeal: 
have a data class `BomRef`. `BomRef` has a property value of type `Optional[str]`.

Comment (jkowalleck) [reply cyclonedx/model/component.py:769]: what d you think, @madpah ?

Comment (madpah) [reply cyclonedx/model/component.py:769]: #170 raised and will consider for this release @jkowalleck

File: cyclonedx/model/bom_ref.py
Comment (jkowalleck) [inline cyclonedx/model/bom_ref.py:52]: a `__str__()` shoult be implemented,
as `BomRef` objets were cast to string, often.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Pull request merged**
